sdk/java/pulumi/src/main/java/com/pulumi/deployment/internal/ReadOrRegisterResource.java

```diff
@@ -0,0 +1,14 @@
+package com.pulumi.deployment.internal;
+
+import com.pulumi.resources.Resource;
+import com.pulumi.resources.ResourceArgs;
+import com.pulumi.resources.ResourceOptions;
+
+import java.util.function.Function;
+
+public interface ReadOrRegisterResource {
```

Comment:
adding targeted interfaces helps with the overall architecture by keeping coupling at minimum

Reply:
I'm confused where this is going. 

The architecture defines a ResourceMonitor service in gRPC. The "real" implementation is hosted inside `pulumi CLI`.

```
// ResourceMonitor is the interface a source uses to talk back to the planning monitor orchestrating the execution.
service ResourceMonitor {
    rpc SupportsFeature(SupportsFeatureRequest) returns (SupportsFeatureResponse) {}
    rpc Invoke(ResourceInvokeRequest) returns (InvokeResponse) {}
    rpc StreamInvoke(ResourceInvokeRequest) returns (stream InvokeResponse) {}
    rpc Call(CallRequest) returns (CallResponse) {}
    rpc ReadResource(ReadResourceRequest) returns (ReadResourceResponse) {}
    rpc RegisterResource(RegisterResourceRequest) returns (RegisterResourceResponse) {}
    rpc RegisterResourceOutputs(RegisterResourceOutputsRequest) returns (google.protobuf.Empty) {}
}
```

Is it still true that users are not expected to interact with it?

My high-level expectation is that in the context of testing against mocks, users implement the Mocks interface and Java SDK uses a test double ResourceMonitor that calls the users's mocks when implementing ReadResource or RegisterResource or Invoke/etc, and does something reasonable for everything else.

Reply:
this is internal interface, used for granular dependency management and to allow loose coupling

Reply:
Ah ok then; I don't have such a strong opinion on the internals. If this is not for the user, that's OK. Bear in mind that all ResourceMonitor methods are conceptually coupled, `RegisterResource` and `ReadResource` work against the same state machine, they're not independent, but you probaly know that. Whatever works best.